mixer/pkg/runtime2/handler/safe.go

```diff
@@ -0,0 +1,120 @@
+// Copyright 2018 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package handler
+
+import (
+	"context"
+	"fmt"
+
+	"istio.io/istio/mixer/pkg/adapter"
+)
+
+// safeHandlerBuilder is a wrapper around adapter.Builder and provides panic protection.
+type safeHandlerBuilder struct {
+	b adapter.HandlerBuilder
+}
+
+// SafeHandler is a wrapper around adapter.Handler and provides panic protection.
+type SafeHandler struct {
+	h adapter.Handler
+}
+
+// SetAdapterConfig calls the same-named method on the wrapped builder. Returns error if the call panics.
+func (s safeHandlerBuilder) SetAdapterConfig(c adapter.Config) (err error) {
+	// Try to detect panic, even if panic was called with nil.
+	reachedEnd := false
+	defer func() {
+		if reachedEnd {
+			return
+		}
+
+		r := recover()
+		err = fmt.Errorf("panic during HandlerBuilder.SetAdapter: '%v' ", r)
+	}()
+
+	s.b.SetAdapterConfig(c)
+	err = nil
+	reachedEnd = true
+	return
+}
+
+// Validate calls the same-named method on the wrapped builder. Returns either the result of the call and nil
+// error, or nil adapter.ConfigError and error that is the result of a panic.
+func (s safeHandlerBuilder) Validate() (cerr *adapter.ConfigErrors, err error) {
```

Comment:
Would it be useful to have a single utility function that can be shared to implement these other functions? Something like:

func safecall(fn func(), name string) error {}

This would invoke the given function pointer with panic protection and return an error using the given name upon failure. This requires a closure on the caller's side, but all this is config time stuff so that should be fine. We just shouldn't use this wrapper for the HandleXXX calls.

Reply:
+1 for doing a utility function. 
Currently all the HandlerXXXX calls are going via generated code. We can either make the generated code panic safe or continue to do what we are doing in current pkg/runtime using a wrapper function for doing [SafeDispatch](https://github.com/istio/istio/blob/master/mixer/pkg/runtime/dispatcher.go#L390) when invoking anything in generated code (I suppose that has closure creating concern). So, we can go ahead with making generated code panic safe.